xla/tests/dot_operation_test.cc

```diff
@@ -41,6 +41,118 @@ limitations under the License.
 namespace xla {
 namespace {
 
+// won't be needed in C++20
+inline bool starts_with(const std::string& a, const std::string& b) {
+  return a.substr(0,b.size()) == b;
+}
+
+class GradFlagVerifier
+{
+  size_t size_x_=0, size_y_=0, flags_=0;
+  bool trivial_ = false;
+public:
+  GradFlagVerifier(XlaBuilder& builder, XlaOp& x, XlaOp& y, PrecisionConfig& cfg) {
+    cfg = xla::PrecisionConfigDEFAULT();
+#if TENSORFLOW_USE_ROCM
+    flags_ = rand() & 3;
+    tsl::setenv("TF_ROCM_PRINT_GRAD_FLAGS", "1", true);
+    cfg.add_operand_precision((flags_ & 1) ? PrecisionConfig::GRADIENT : PrecisionConfig::ACTIVATION);
+    cfg.add_operand_precision((flags_ & 2) ? PrecisionConfig::GRADIENT : PrecisionConfig::ACTIVATION);
+
+    Shape x_shape = *(builder.GetShape(x));
+    if(x_shape.is_static()) {
+      size_x_ = 1;
+      int nontrivial_dims = 0;
+      int nd=0;
+      for(auto d: x_shape.dimensions()) {
+        // If one of the inputs has no dimensions larger than 1, GEMM call may be optimized out
+        if(d > 1)
+          nontrivial_dims++;
+        // For F16 and BF16, last 2 dimensions of both inputs will be padded to multiples of 8
+        if(nd+2>=x_shape.rank() && (x_shape.element_type()==F16 || x_shape.element_type()==BF16))
+          d = (d+7) & ~7;
+        size_x_ *= d;
+        nd++;
+      }
+      if(nontrivial_dims <= 1)
+        trivial_ = true;
+      size_x_ *= xla::primitive_util::BitWidth(x_shape.element_type())/8;
+    }
+    Shape y_shape = *(builder.GetShape(y));
+    if(y_shape.is_static()) {
+      size_y_ = 1;
+      int nontrivial_dims = 0;
+      int nd = 0;
+      for(auto d: y_shape.dimensions()) {
+        if(d > 1)
+          nontrivial_dims++;
+
+        if(nd+2 >= y_shape.rank() && (y_shape.element_type()==F16 || y_shape.element_type()==BF16))
+          d = (d+7) & ~7;
+        size_y_ *= d;
+        nd++;
+      }
+      if(nontrivial_dims <= 1)
+        trivial_ = true;
+      size_y_ *= xla::primitive_util::BitWidth(y_shape.element_type())/8;
+    }
+    //printf("Shapes: %s, %s\n", x_shape.ToString().c_str(), y_shape.ToString().c_str());
```

Comment:
Does it make sense to remove these?

Reply:
Absolutely, once it's no longer marked as .work in progress